File: fastapi/app/schemas/place_dna.py
from typing import Literal
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field, StrictFloat, StrictInt, StrictStr

Rarity = Literal["Common", "Rare", "Epic", "Legendary", "Vulnerable"]


class PlaceDNAStats(BaseModel):
    model_config = ConfigDict(extra="forbid")

    green_cover: StrictInt = Field(ge=0, le=100)
    population_pressure: StrictInt = Field(ge=0, le=100)
    built_up_density: StrictInt = Field(ge=0, le=100)
    water_access: StrictInt = Field(ge=0, le=100)
    connectivity: StrictInt = Field(ge=0, le=100)
    liveability: StrictInt = Field(ge=0, le=100)


class PlaceDNALandmark(BaseModel):
    model_config = ConfigDict(extra="forbid")

    name: StrictStr
    distance_m: float | None = Field(default=None, ge=0)
    image_url: StrictStr | None = None
    source: StrictStr | None = None
    osm_type: StrictStr | None = None
    osm_id: StrictInt | None = None
    wikidata_id: StrictStr | None = None


class PlaceDNALocation(BaseModel):
    model_config = ConfigDict(extra="forbid")

    lat: StrictFloat = Field(ge=-90, le=90)
    lon: StrictFloat = Field(ge=-180, le=180)
    radius_m: StrictInt = Field(ge=100, le=2000)


class PlaceDNAResponse(BaseModel):
    model_config = ConfigDict(extra="forbid")

    id: UUID | None = None
    place_name: StrictStr
    title: StrictStr
    rarity: Rarity
    region_type: StrictStr
    stats: PlaceDNAStats
    traits: list[StrictStr]
    description: StrictStr
    landmark: PlaceDNALandmark
    location: PlaceDNALocation

File: nextjs/src/types/placedna.ts
import type { LucideIcon } from "lucide-react";

export type Rarity =
  | "Common"
  | "Rare"
  | "Epic"
  | "Legendary"
  | "Vulnerable";

export type DecorTone =
  | "neutral"
  | "accent"
  | "secondary"
  | "tertiary"
  | "quaternary";

export type StatTone =
  | "neutral"
  | "success"
  | "info"
  | "secondary"
  | "warning"
  | "error";

export type PlaceStat = {
  label: string;
  value: number;
  tone?: StatTone;
};

export type PlaceCardPreviewData = {
  title: string;
  rarity: Rarity;
  regionType: string;
  landmarkName: string;
  coordinates: string;
  stats: PlaceStat[];
  traits: string[];
};

export type PlaceDNAStats = {
  green_cover: number;
  population_pressure: number;
  built_up_density: number;
  water_access: number;
  connectivity: number;
  liveability: number;
};

export type PlaceDNALandmark = {
  name: string;
  distance_m: number | null;
  image_url: string | null;
};

export type PlaceDNALocation = {
  lat: number;
  lon: number;
  radius_m: number;
};

export type PlaceDNAResponse = {
  id: string;
  place_name: string;
  title: string;
  rarity: Rarity;
  region_type: string;
  stats: PlaceDNAStats;
  traits: string[];
  description: string;
  landmark: PlaceDNALandmark;
  location: PlaceDNALocation;
};

export type SelectedMapLocation = {
  lat: number;
  lon: number;
};

export type MapFocusRequest = {
  id: number;
  location: SelectedMapLocation;
};

export type NavLink = {
  label: string;
  href: string;
};

export type HeroPill = {
  label: string;
  tone: DecorTone;
};

export type ContentCard = {
  title: string;
  description: string;
  icon: LucideIcon;
  tone: DecorTone;
  label?: string;
};

export type RarityTier = {
  name: Rarity;
  description: string;
  signal: string;
  icon: LucideIcon;
  tone: DecorTone;
  spotlight?: boolean;
};

export type DemoStage = {
  title: string;
  description: string;
  tone: DecorTone;
};
